File: content/problems/0229-majority-element-ii/solutions/boyer-moore-two-candidates/Solution.java
class Solution {
    public List<Integer> majorityElement(int[] nums){
        int first=0,second=1,count1=0,count2=0;
        for(int value:nums){
            if(value==first)count1++;
            else if(value==second)count2++;
            else if(count1==0){
                first=value;
                count1=1;
            }else if(count2==0){
                second=value;
                count2=1;
            }else{
                count1--;
                count2--;
            }
        }
        count1=count2=0;
        for(int value:nums){
            if(value==first)count1++;
            else if(value==second)count2++;
        }
        List<Integer> answer=new ArrayList<>();
        if(count1>nums.length/3)answer.add(first);
        if(count2>nums.length/3)answer.add(second);
        Collections.sort(answer);
        return answer;
    }
}

File: content/problems/0229-majority-element-ii/solutions/boyer-moore-two-candidates/Solution.kt
class Solution {
    fun majorityElement(nums:IntArray):List<Int>{
        var first=0;
        var second=1;
        var count1=0;
        var count2=0;
        for(value in nums){
            when{
                value==first->count1++;
                value==second->count2++;
                count1==0->{
                    first=value;
                    count1=1
                };
                count2==0->{
                    second=value;
                    count2=1
                };
                else->{
                    count1--;
                    count2--
                }
            }
        };
        count1=nums.count{
            it==first
        };
        count2=nums.count{
            it==second
        };
        val answer=mutableListOf<Int>();
        if(count1>nums.size/3)answer.add(first);
        if(count2>nums.size/3)answer.add(second);
        answer.sort();
        return answer
    }
}
